src/snowflake/snowpark/mock/_functions.py

```diff
@@ -933,6 +933,45 @@ def mock_to_variant(expr: ColumnEmulator):
     return res
 
 
+def _compare(x: Any, y: Any) -> Tuple[Any, Any]:
+    """
+    Compares two values based on the rules described for greatest/least
+    https://docs.snowflake.com/en/sql-reference/functions/least#usage-notes
+    """
+    if x is None or y is None:
+        return (None, None)
+
+    _x = x if isinstance(x, Number) else float(x)
+    _y = y if isinstance(y, Number) else float(y)
+
+    if _x > _y:
+        return (_x, _y)
+    else:
+        return (_y, _x)
```

Comment:
do we consider non-numeric types here?
I see in the description https://docs.snowflake.com/en/sql-reference/functions/least:

`LEAST supports all data types, including VARIANT.`

Reply:
I can add some more test cases to be sure. From what I can tell anything non numeric gets cast to a numeric type.

Reply:
interesting, if I compare a string "abc" with a datetime "2024-01-1", both would be cast to numeric first?

Reply:
That case actually generates an exception:
```
SELECT greatest("A", "B") FROM ( SELECT "A", to_timestamp("B") AS "B" FROM ( SELECT $1 AS "A", $2 AS "B" FROM  VALUES ('abc' :: STRING, '2024-01-01 00:00:00' :: STRING)));
```
```
Timestamp 'abc' is not recognized
```

Reply:
Although if thy are all strings it does look like they aren't converted to numerics so the _compare function is incorrect. I'll fix it up.

Reply:
thanks! do you know what exception class it's raising?
I'm thinking of raising the same error to align the error experience in live mode.
for example, I noticed that in the code `type(_x)(y)`, this might raise `ValueError`, I'm thinking if we should wrap this into a snowpark client exception